### PR TITLE
Hardware inventory

### PR DESCRIPTION
**Pull request by julien-c**
I hesitated putting this one in `@huggingface/tasks`, or creating a new `@huggingface/hardware`.

What do you think?

The idea is the community to be able to contribute to that list.

### What I picked for now ⤵️

Because I'm lazy – and because it's somewhat linked to https://github.com/huggingface/huggingface.js/pull/659 – i've added it to `tasks`.

### CODE REVIEW

**Review by coyotte508 — APPROVED** [review]

**Review by krampstudio — APPROVED** [review]
:+1:

**Review by enzostvs — APPROVED** [review]

File: packages/tasks/src/hardware.ts [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+			},
+			"RTX 3090 Ti": {
+				tflops: 7.57,
+				memory: [24],
+			},
+			"RTX 3080 Ti": {
+				tflops: 7.57,
+				memory: [24, 16],
+			},
+			"RTX 3080": {
```
Comment (osanseviero) [inline packages/tasks/src/hardware.ts:103]: This has 12 only I think? (mine has 12 at least)

File: packages/tasks/src/hardware.ts [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+ */
+export const TFLOPS_THRESHOLD_WHITE_HOUSE_MODEL_TRAINING_TOTAL = 10 ** 14;
+export const TFLOPS_THRESHOLD_WHITE_HOUSE_MODEL_TRAINING_TOTAL_BIOLOGY = 10 ** 11;
+export const TFLOPS_THRESHOLD_WHITE_HOUSE_CLUSTER = 10 ** 8;
+
+/**
+ * EU AI Act
+ * https://ec.europa.eu/commission/presscorner/detail/en/qanda_21_1683
+ */
+export const TFLOPS_THRESHOLD_EU_AI_ACT_MODEL_TRAINING_TOTAL = 10 ** 13;
```
Comment (osanseviero) [inline packages/tasks/src/hardware.ts:13]: This part feels a bit weird here, at least without context of how it's used

File: packages/tasks/src/hardware.ts [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+	 */
+	tflops: number;
+	/**
+	 * If an array is specified, options of memory size (can be VRAM, unified RAM)
+	 * e.g. an A100 exists in 40 or 80 GB.
+	 */
+	memory?: number[];
+}
+
+export const DEFAULT_MEMORY_OPTIONS = [8, 16, 24, 32, 40, 48, 64, 80];
```
Comment (osanseviero) [inline packages/tasks/src/hardware.ts:34]: Also 12?

File: packages/tasks/src/hardware.ts [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+				tflops: 8.58,
+				memory: [24, 16],
+			},
+			"RTX 4080 SUPER": {
+				tflops: 8.58,
+				memory: [16],
+			},
+			"RTX 4080": {
+				tflops: 8.58,
+				memory: [12],
```
Comment (pcuenca) [inline packages/tasks/src/hardware.ts:73]:
```suggestion
				tflops: 48.7,
				memory: [16],
```

Source: https://www.tomshardware.com/pc-components/gpus/nvidia-geforce-rtx-4080-super-review

File: packages/tasks/src/hardware.ts [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+			},
+			"RTX 4070": {
+				tflops: 8.58,
+				memory: [12, 8],
+			},
```
Comment (pcuenca) [inline packages/tasks/src/hardware.ts:78]:
```suggestion
			"RTX 4070": {
				tflops: 29.15,
				memory: [12],
			},
```

Source: https://www.techpowerup.com/gpu-specs/geforce-rtx-4070.c3924

File: packages/tasks/src/hardware.ts [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+			"RTX 4070 Ti": {
+				tflops: 8.58,
```
Comment (pcuenca) [inline packages/tasks/src/hardware.ts:80]:
```suggestion
				tflops: 40.09,
```

Source: https://www.techpowerup.com/gpu-specs/geforce-rtx-4070-ti.c3950

File: packages/tasks/src/hardware.ts [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+				memory: [12],
+			},
+			"RTX 4070 Super": {
+				tflops: 8.58,
```
Comment (pcuenca) [inline packages/tasks/src/hardware.ts:84]:
```suggestion
				tflops: 35.48,
```

Source: https://www.techpowerup.com/gpu-specs/geforce-rtx-4070-super.c4186, https://www.tomshardware.com/pc-components/gpus/nvidia-geforce-rtx-4070-super-review-boosted-clocks-and-core-counts-for-the-same-dollar599-as-the-vanilla-4070

File: packages/tasks/src/hardware.ts [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+				memory: [12],
+			},
+			"RTX 4070 Ti Super": {
+				tflops: 7.57,
```
Comment (pcuenca) [inline packages/tasks/src/hardware.ts:88]:
```suggestion
				tflops: 44.1,
```

Source: https://videocardz.com/nvidia/geforce-40/geforce-rtx-4070-ti-super

File: packages/tasks/src/hardware.ts [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+				memory: [24],
+			},
+			"RTX 3090 Ti": {
+				tflops: 7.57,
+				memory: [24],
+			},
+			"RTX 3080 Ti": {
+				tflops: 7.57,
+				memory: [24, 16],
```
Comment (pcuenca) [inline packages/tasks/src/hardware.ts:101]:
```suggestion
				tflops: 34.1,
				memory: [12],
```

Source: https://www.anandtech.com/show/17204/nvidia-quietly-launches-geforce-rtx-3080-12gb-more-vram-more-power-more-money

File: packages/tasks/src/hardware.ts (hunk 1 above)
Comment (pcuenca) [reply packages/tasks/src/hardware.ts:103]: That's what I thought, there's apparently a version with 10 GB as well

File: packages/tasks/src/hardware.ts [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+			},
+			"RTX 3080": {
+				tflops: 7.57,
+				memory: [24, 16, 8],
```
Comment (pcuenca) [inline packages/tasks/src/hardware.ts:105]:
```suggestion
				tflops: 30.6,
				memory: [12, 10],
```

File: packages/tasks/src/hardware.ts [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+			"RTX 4070 Super": {
+				tflops: 8.58,
+				memory: [12],
+			},
+			"RTX 4070 Ti Super": {
+				tflops: 7.57,
+				memory: [16],
+			},
+			"RTX 3090": {
+				tflops: 7.57,
```
Comment (pcuenca) [inline packages/tasks/src/hardware.ts:92]:
```suggestion
				tflops: 35.58,
```

Source: https://news.ycombinator.com/item?id=34261797#:~:text=The%20RTX%203090%20has%2035.58,http...%20%7C%20Hacker%20News&text=johndough%20on%20Jan%205%2C%202023,performance%20on%20a%20sin...

File: packages/tasks/src/hardware.ts [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+				memory: [24],
+			},
+			"RTX 3090 Ti": {
+				tflops: 7.57,
```
Comment (pcuenca) [inline packages/tasks/src/hardware.ts:96]:
```suggestion
				tflops: 40,
```

Source: https://www.theverge.com/2022/3/29/23001134/nvidia-rtx-3090-ti-specs-release-date-pricing

File: packages/tasks/src/hardware.ts [hunk 13]
```diff
@@ -0,0 +1,305 @@
...
+				tflops: 13.6,
+				memory: [16, 24, 32, 64, 96, 128],
+			},
+			"Apple M2 Max": {
+				tflops: 13.49,
+				memory: [16, 24, 32, 64, 96, 128],
+			},
+			"Apple M2 Ultra": {
+				tflops: 27.2,
+				memory: [16, 24, 32, 64, 96, 128],
```
Comment (pcuenca) [inline packages/tasks/src/hardware.ts:287]:
```suggestion
				memory: [64, 96, 128, 192],
```

File: packages/tasks/src/hardware.ts [hunk 14]
```diff
@@ -0,0 +1,305 @@
...
+	},
+	"Apple Silicon": {
+		"-": {
+			"Apple M1": {
+				tflops: 2.6,
+				memory: [16, 24, 32, 64, 96, 128],
+			},
+			"Apple M1 Pro": {
+				tflops: 5.2,
+				memory: [16, 24, 32, 64, 96, 128],
```
Comment (pcuenca) [inline packages/tasks/src/hardware.ts:263]:
```suggestion
				memory: [16, 24, 32],
```

Not sure ever was a 24 GB configuration, but max was announced to be 32: https://www.apple.com/hu/newsroom/2021/10/introducing-m1-pro-and-m1-max-the-most-powerful-chips-apple-has-ever-built/

File: packages/tasks/src/hardware.ts [hunk 15]
```diff
@@ -0,0 +1,305 @@
...
+			},
+			"Apple M1 Max": {
+				tflops: 10.4,
+				memory: [16, 24, 32, 64, 96, 128],
```
Comment (pcuenca) [inline packages/tasks/src/hardware.ts:267]:
```suggestion
				memory: [16, 24, 32, 64],
```

File: packages/tasks/src/hardware.ts [hunk 16]
```diff
@@ -0,0 +1,305 @@
...
+			"Ryzen Zen 2  3000 (Ryzen 3)": {
+				tflops: 0.72,
+			},
+		},
+	},
+	"Apple Silicon": {
+		"-": {
+			"Apple M1": {
+				tflops: 2.6,
+				memory: [16, 24, 32, 64, 96, 128],
```
Comment (pcuenca) [inline packages/tasks/src/hardware.ts:259]:
```suggestion
				memory: [8, 16],
```

I think

File: packages/tasks/src/hardware.ts [hunk 17]
```diff
@@ -0,0 +1,305 @@
...
+				tflops: 10.4,
+				memory: [16, 24, 32, 64, 96, 128],
+			},
+			"Apple M1 Ultra": {
+				tflops: 21,
+				memory: [16, 24, 32, 64, 96, 128],
+			},
+			"Apple M2": {
+				tflops: 3.6,
+				memory: [16, 24, 32, 64, 96, 128],
```
Comment (pcuenca) [inline packages/tasks/src/hardware.ts:275]:
```suggestion
				memory: [8, 16, 24],
```

Source: https://en.wikipedia.org/wiki/Apple_M2

File: packages/tasks/src/hardware.ts [hunk 18]
```diff
@@ -0,0 +1,305 @@
...
+			},
+			"Apple M2 Pro": {
+				tflops: 13.6,
+				memory: [16, 24, 32, 64, 96, 128],
```
Comment (pcuenca) [inline packages/tasks/src/hardware.ts:279]:
```suggestion
				memory: [16, 24, 32],
```

File: packages/tasks/src/hardware.ts [hunk 19]
```diff
@@ -0,0 +1,305 @@
...
+			},
+			"Apple M2 Max": {
+				tflops: 13.49,
+				memory: [16, 24, 32, 64, 96, 128],
```
Comment (pcuenca) [inline packages/tasks/src/hardware.ts:283]:
```suggestion
				memory: [32, 64, 96],
```

File: packages/tasks/src/hardware.ts [hunk 20]
```diff
@@ -0,0 +1,305 @@
...
+			},
+			"Apple M2 Ultra": {
+				tflops: 27.2,
+				memory: [16, 24, 32, 64, 96, 128],
+			},
+			"Apple M3": {
+				tflops: 2.84,
+				memory: [16, 24, 32, 64, 96, 128],
```
Comment (pcuenca) [inline packages/tasks/src/hardware.ts:291]:
```suggestion
				memory: [8, 16, 24],
```

File: packages/tasks/src/hardware.ts [hunk 21]
```diff
@@ -0,0 +1,305 @@
...
+			},
+			"Apple M3 Pro": {
+				tflops: 14,
+				memory: [16, 24, 32, 64, 96, 128],
```
Comment (pcuenca) [inline packages/tasks/src/hardware.ts:295]:
```suggestion
				memory: [18, 36],
```

Source: Apple purchase page

File: packages/tasks/src/hardware.ts [hunk 22]
```diff
@@ -0,0 +1,305 @@
...
+			},
+			"Apple M3 Max": {
+				tflops: 14.2,
+				memory: [16, 24, 32, 64, 96, 128],
```
Comment (pcuenca) [inline packages/tasks/src/hardware.ts:299]:
```suggestion
				memory: [36, 48, 64, 96, 128],
```

File: packages/tasks/src/hardware.ts (hunk 3 above)
Comment (pcuenca) [reply packages/tasks/src/hardware.ts:34]: What is it used for? Missing from GPU/CPU: 10, 12, 18, 20, 36, 96, 128, 192

**Review by pcuenca — COMMENTED** [review]
Verified the nvidia GPUs and the RAM in the Apple chips, ignored all the rest.

**Comment by rwightman** [comment]
If things are being simplified such that one flops number is used, it's probably good to focus on the most used/usable option, i.e. for Nvidia cards the non-sparse half-precision tensor core FLOPS is what most people will be operating under for train workloads, many inferencing ... although inferencing is starting to see heavier use of 8-bit int or float precisions... 

Comparing any Nvidia datacenter GPU on the base (non tensorcore) FP32 flops does not make any sense. Nobody uses them in that mode, they cut the capability of that mode from gamer GPUs so that they can fit more TC on there, maybe a few will be using TF32 on the TC, but most will be in mixed or half precision for matrix multiplies....

**Comment by rwightman** [comment]
Also some of the latest Intel CPU (and ARM) have bfloat16 and/or int8 instruction sets for mixed/half precision training and inference, so that's something to watch. Up to this point there were usually just float32 flops to consider on CPUs.

**Comment by apolinario** [comment]
An example of what @rwightman is bringing up is an A100 
image

image

A100s have only 19.49 TFLOPS on fp32 - which would make it look weaker than A10, A10G, L4 and L40s which clearly it isn't in real world use-cases e.g.:
image

So I agree with Ross that maybe best to use fp16 as baseline flops vs fp32 for all GPUs. I think it may correlate better with perceived real world performance

File: packages/tasks/src/hardware.ts (hunk 2 above)
Comment (julien-c) [reply packages/tasks/src/hardware.ts:13]: haha yes it's mostly for fun, if we want to add a `x-risk` label on GPU rich users' profiles at some point =)

File: packages/tasks/src/hardware.ts (hunk 3 above)
Comment (julien-c) [reply packages/tasks/src/hardware.ts:34]: it's used as a default list to pick from when `SKUS` does not define a `memory` list of options. so mostly for CPUs!

Comment (julien-c) [reply packages/tasks/src/hardware.ts:34]: in light of this new info, what new list items do you suggest? (i think it's ok if it's not 100% exhaustive, people can choose the closest one)

Comment (julien-c) [reply packages/tasks/src/hardware.ts]: eff64a62729620bc089a2ca4f2d07a74f7b72e1f

**Comment by julien-c** [comment]
@apolinario @rwightman, sounds good, where can one find the values for FP16?

**Comment by julien-c** [comment]
i've switched from FP32 => FP16 for NV and AMD GPUs.

TFlops for CPUs are still kinda random, if anyone wants to give me a hand?

**Comment by apolinario** [comment]
Regarding CPUs, the same generation of processor can vary in TFLOPS, I think this website can provide decent directions (https://www.cpu-monkey.com/en/benchmark-intel_core_i5_13600k-bench_11) and I'm happy to help with data-entry for CPUs

However, wondering how to handle the scenario: 
- i5 11th generation (Intel Core i5-11320H) can have 2 TFLOPS (https://www.cpu-monkey.com/en/benchmark-intel_core_i5_11320h-bench_11)
- i5 11th generation (Intel Core i5-11400) can have 0.5 TFLOPS (https://www.cpu-monkey.com/en/benchmark-intel_core_i5_11400-bench_11) 

Suggestion: having for each processor a `low end` and `high end` (so we don't need to enter every variant) 
```
"Intel Core 11th Generation (i5) - low end": {
    tflops: 0.5,
},
"Intel Core 11th Generation (i5) - high end": {
    tflops: 0.5,
},
```
Would be bad if the user doesn't know how low/high end is their cpu, but if they don't know if theirs is low or high end they may not know the specific to pick either 🤔 - if you think this is a good way happy to fill in the data for this

**Comment by julien-c** [comment]
honestly i would just suggest inputting a value in the middle of `low end` and `high end`

Especially given this comment line here at the top of `export interface HardwareSpec`:
```ts
 This is only approximate/theoretical and shouldn't be taken too seriously.
```

File: packages/tasks/src/hardware.ts [hunk 23]
```diff
@@ -0,0 +1,309 @@
...
+				tflops: 1,
+			},
+			"Intel Core 11th Generation (i7)": {
+				tflops: 0.76,
+			},
+			"Intel Core 11th Generation (i5)": {
+				tflops: 0.76,
+			},
+			"Intel Core 11th Generation (i3)": {
+				tflops: 0.76,
```
Comment (apolinario) [inline packages/tasks/src/hardware.ts:196]:
```suggestion
				tflops: 0.35,
```
https://www.cpu-monkey.com/en/cpu-intel_core_i3_11100b

File: packages/tasks/src/hardware.ts [hunk 24]
```diff
@@ -0,0 +1,309 @@
...
+				tflops: 1,
+			},
+			"Intel Core 11th Generation (i9)": {
+				tflops: 1,
+			},
+			"Intel Core 11th Generation (i7)": {
+				tflops: 0.76,
+			},
+			"Intel Core 11th Generation (i5)": {
+				tflops: 0.76,
```
Comment (apolinario) [inline packages/tasks/src/hardware.ts:193]:
```suggestion
				tflops: 0.5,
```
average between 
11400T (lower end) and 11600k (higher end) excluding ultra-high end low sales version

File: packages/tasks/src/hardware.ts [hunk 25]
```diff
@@ -0,0 +1,309 @@
...
+				tflops: 2,
+			},
+			"Intel Core 12th Generation (i3)": {
+				tflops: 1,
+			},
+			"Intel Core 11th Generation (i9)": {
+				tflops: 1,
+			},
+			"Intel Core 11th Generation (i7)": {
+				tflops: 0.76,
```
Comment (apolinario) [inline packages/tasks/src/hardware.ts:190]:
```suggestion
				tflops: 0.6,
```

File: packages/tasks/src/hardware.ts [hunk 26]
```diff
@@ -0,0 +1,309 @@
...
+				tflops: 2,
+			},
+			"Intel Core 12th Generation (i5)": {
+				tflops: 2,
+			},
+			"Intel Core 12th Generation (i3)": {
+				tflops: 1,
+			},
+			"Intel Core 11th Generation (i9)": {
+				tflops: 1,
```
Comment (apolinario) [inline packages/tasks/src/hardware.ts:187]:
```suggestion
				tflops: 0.7,
```

File: packages/tasks/src/hardware.ts [hunk 27]
```diff
@@ -0,0 +1,309 @@
...
+				tflops: 0.76,
+			},
+			"Intel Core 10th Generation (i7)": {
+				tflops: 0.76,
+			},
+			"Intel Core 10th Generation (i5)": {
+				tflops: 0.76,
+			},
+			"Intel Core 10th Generation (i3)": {
+				tflops: 0.76,
```
Comment (apolinario) [inline packages/tasks/src/hardware.ts:208]:
```suggestion
				tflops: 0.44,
```

File: packages/tasks/src/hardware.ts [hunk 28]
```diff
@@ -0,0 +1,309 @@
...
+			},
+			"Intel Core 10th Generation (i9)": {
+				tflops: 0.76,
+			},
+			"Intel Core 10th Generation (i7)": {
+				tflops: 0.76,
+			},
+			"Intel Core 10th Generation (i5)": {
+				tflops: 0.76,
```
Comment (apolinario) [inline packages/tasks/src/hardware.ts:205]:
```suggestion
				tflops: 0.46,
```

File: packages/tasks/src/hardware.ts [hunk 29]
```diff
@@ -0,0 +1,309 @@
...
+			},
+			"Intel Core 11th Generation (i3)": {
+				tflops: 0.76,
+			},
+			"Intel Core 10th Generation (i9)": {
+				tflops: 0.76,
+			},
+			"Intel Core 10th Generation (i7)": {
+				tflops: 0.76,
```
Comment (apolinario) [inline packages/tasks/src/hardware.ts:202]:
```suggestion
				tflops: 0.46,
```

File: packages/tasks/src/hardware.ts [hunk 30]
```diff
@@ -0,0 +1,309 @@
...
+			},
+			"Intel Core 11th Generation (i5)": {
+				tflops: 0.76,
+			},
+			"Intel Core 11th Generation (i3)": {
+				tflops: 0.76,
+			},
+			"Intel Core 10th Generation (i9)": {
+				tflops: 0.76,
```
Comment (apolinario) [inline packages/tasks/src/hardware.ts:199]:
```suggestion
				tflops: 0.46,
```

File: packages/tasks/src/hardware.ts [hunk 31]
```diff
@@ -0,0 +1,309 @@
...
+				tflops: 2,
+			},
+			"Intel Core 12th Generation (i7)": {
+				tflops: 2,
+			},
+			"Intel Core 12th Generation (i5)": {
+				tflops: 2,
+			},
+			"Intel Core 12th Generation (i3)": {
+				tflops: 1,
```
Comment (apolinario) [inline packages/tasks/src/hardware.ts:184]:
```suggestion
				tflops: 0.53,
```

File: packages/tasks/src/hardware.ts [hunk 32]
```diff
@@ -0,0 +1,309 @@
...
+				tflops: 2,
+			},
+			"Intel Core 12th Generation (i9)": {
+				tflops: 2,
+			},
+			"Intel Core 12th Generation (i7)": {
+				tflops: 2,
+			},
+			"Intel Core 12th Generation (i5)": {
+				tflops: 2,
```
Comment (apolinario) [inline packages/tasks/src/hardware.ts:181]:
```suggestion
				tflops: 0.65,
```

File: packages/tasks/src/hardware.ts [hunk 33]
```diff
@@ -0,0 +1,309 @@
...
+				tflops: 3,
+			},
+			"Intel Core 13th Generation (i3)": {
+				tflops: 2,
+			},
+			"Intel Core 12th Generation (i9)": {
+				tflops: 2,
+			},
+			"Intel Core 12th Generation (i7)": {
+				tflops: 2,
```
Comment (apolinario) [inline packages/tasks/src/hardware.ts:178]:
```suggestion
				tflops: 0.77,
```

File: packages/tasks/src/hardware.ts [hunk 34]
```diff
@@ -0,0 +1,309 @@
...
+				tflops: 3.8,
+			},
+			"Intel Core 13th Generation (i7)": {
+				tflops: 3.6,
+			},
+			"Intel Core 13th Generation (i5)": {
+				tflops: 3,
+			},
+			"Intel Core 13th Generation (i3)": {
+				tflops: 2,
```
Comment (apolinario) [inline packages/tasks/src/hardware.ts:172]:
```suggestion
				tflops: 0.57,
```

File: packages/tasks/src/hardware.ts [hunk 35]
```diff
@@ -0,0 +1,309 @@
...
+				tflops: 3,
+			},
+			"Intel Core 13th Generation (i9)": {
+				tflops: 3.8,
+			},
+			"Intel Core 13th Generation (i7)": {
+				tflops: 3.6,
+			},
+			"Intel Core 13th Generation (i5)": {
+				tflops: 3,
```
Comment (apolinario) [inline packages/tasks/src/hardware.ts:169]:
```suggestion
				tflops: 0.68,
```

File: packages/tasks/src/hardware.ts [hunk 36]
```diff
@@ -0,0 +1,309 @@
...
+				tflops: 3.6,
+			},
+			"Xeon 2th Generation (Cascade Lake)": {
+				tflops: 3,
+			},
+			"Intel Core 13th Generation (i9)": {
+				tflops: 3.8,
+			},
+			"Intel Core 13th Generation (i7)": {
+				tflops: 3.6,
```
Comment (apolinario) [inline packages/tasks/src/hardware.ts:166]:
```suggestion
				tflops: 0.82,
```

File: packages/tasks/src/hardware.ts [hunk 37]
```diff
@@ -0,0 +1,309 @@
...
+				tflops: 3.8,
+			},
+			"Xeon 3th Generation (Ice Lake)": {
+				tflops: 3.6,
+			},
+			"Xeon 2th Generation (Cascade Lake)": {
+				tflops: 3,
+			},
+			"Intel Core 13th Generation (i9)": {
+				tflops: 3.8,
```
Comment (apolinario) [inline packages/tasks/src/hardware.ts:163]:
```suggestion
				tflops: 0.85,
```

File: packages/tasks/src/hardware.ts [hunk 38]
```diff
@@ -0,0 +1,309 @@
...
+			"RX 7600 XT": {
+				tflops: 45.14,
+				memory: [16, 8],
+			},
+		},
+	},
+	CPU: {
+		Intel: {
+			"Xeon 4th Generation (Sapphire Rapids)": {
+				tflops: 3.8,
```
Comment (apolinario) [inline packages/tasks/src/hardware.ts:154]:
```suggestion
				tflops: 1.3,
```
Sauce: https://www.intel.com/content/dam/support/us/en/documents/processors/APP-for-Intel-Xeon-Processors.pdf

File: packages/tasks/src/hardware.ts [hunk 39]
```diff
@@ -0,0 +1,309 @@
...
+			},
+			"Xeon 3th Generation (Ice Lake)": {
+				tflops: 3.6,
```
Comment (apolinario) [inline packages/tasks/src/hardware.ts:157]:
```suggestion
				tflops: 0.8,
```
Sauce: https://www.intel.com/content/dam/support/us/en/documents/processors/APP-for-Intel-Xeon-Processors.pdf

File: packages/tasks/src/hardware.ts [hunk 40]
```diff
@@ -0,0 +1,309 @@
...
+			},
+			"Xeon 2th Generation (Cascade Lake)": {
+				tflops: 3,
```
Comment (apolinario) [inline packages/tasks/src/hardware.ts:160]:
```suggestion
				tflops: 0.55,
```
Sauce: https://www.intel.com/content/dam/support/us/en/documents/processors/APP-for-Intel-Xeon-Processors.pdf

File: packages/tasks/src/hardware.ts [hunk 41]
```diff
@@ -0,0 +1,309 @@
...
+			"Intel Core 10th Generation (i5)": {
+				tflops: 0.76,
+			},
+			"Intel Core 10th Generation (i3)": {
+				tflops: 0.76,
+			},
+		},
+		AMD: {
+			"EPYC 4th Generation (Genoa)": {
+				tflops: 4,
```
Comment (apolinario) [inline packages/tasks/src/hardware.ts:213]:
```suggestion
				tflops: 5.3,
```
not clear, source: https://www.hpcwire.com/2022/11/10/amds-4th-gen-epyc-genoa-96-5nm-cores-across-12-compute-chiplets/

File: packages/tasks/src/hardware.ts [hunk 42]
```diff
@@ -0,0 +1,309 @@
...
+				tflops: 3,
+			},
+			"EPYC 2th Generation (Rome)": {
+				tflops: 2.9,
+			},
+			"EPYC 1st Generation (Naples)": {
+				tflops: 2.3,
+			},
+			"Ryzen Zen4 7000 (Ryzen 9)": {
+				tflops: 2.3,
```
Comment (apolinario) [inline packages/tasks/src/hardware.ts:225]:
```suggestion
				tflops: 0.96,
```
sauce: https://gadgetversus.com/processor/amd-ryzen-9-5950x-gflops-performance/

File: packages/tasks/src/hardware.ts [hunk 43]
```diff
@@ -0,0 +1,309 @@
...
+			},
+			"Ryzen Zen4 7000 (Ryzen 7)": {
+				tflops: 1.33,
```
Comment (apolinario) [inline packages/tasks/src/hardware.ts:228]:
```suggestion
				tflops: 0.56,
```
https://gadgetversus.com/processor/amd-ryzen-7-pro-7745-vs-amd-ryzen-7-7800x3d/

File: packages/tasks/src/hardware.ts [hunk 44]
```diff
@@ -0,0 +1,309 @@
...
+			},
+			"Ryzen Zen4 7000 (Ryzen 5)": {
+				tflops: 1.33,
```
Comment (apolinario) [inline packages/tasks/src/hardware.ts:231]:
```suggestion
				tflops: 0.56,
```
https://gadgetversus.com/processor/amd-ryzen-5-7500f-vs-amd-ryzen-5-7600x/

File: packages/tasks/src/hardware.ts (hunk 41 above)
Comment (apolinario) [reply packages/tasks/src/hardware.ts:213]: Here it gives 4.6 https://gadgetversus.com/processor/amd-epyc-9654-specs/

File: packages/tasks/src/hardware.ts [hunk 45]
```diff
@@ -0,0 +1,309 @@
...
+			"Intel Core 10th Generation (i3)": {
+				tflops: 0.76,
+			},
+		},
+		AMD: {
+			"EPYC 4th Generation (Genoa)": {
+				tflops: 4,
+			},
+			"EPYC 3th Generation (Milan)": {
+				tflops: 3,
```
Comment (apolinario) [inline packages/tasks/src/hardware.ts:216]:
```suggestion
				tflops: 2.4,
```
source: https://gadgetversus.com/processor/amd-epyc-7543p-specs/

File: packages/tasks/src/hardware.ts [hunk 46]
```diff
@@ -0,0 +1,309 @@
...
+			},
+			"EPYC 2th Generation (Rome)": {
+				tflops: 2.9,
```
Comment (apolinario) [inline packages/tasks/src/hardware.ts:219]:
```suggestion
				tflops: 0.6,
```
source: https://gadgetversus.com/processor/amd-epyc-7232p-specs/

File: packages/tasks/src/hardware.ts [hunk 47]
```diff
@@ -0,0 +1,309 @@
...
+			},
+			"EPYC 1st Generation (Naples)": {
+				tflops: 2.3,
```
Comment (apolinario) [inline packages/tasks/src/hardware.ts:222]:
```suggestion
				tflops: 0.6,
```
source: https://gadgetversus.com/processor/amd-epyc-7371-specs/

**Comment by rwightman** [comment]
> i've switched from FP32 => FP16 for NV and AMD GPUs.
> 
> TFlops for CPUs are still kinda random, if anyone wants to give me a hand?

CPU flops can be quite confusing because there are FPLOPS from the old FPU (floating point unit), there are FLOPS from the  SIMD/AVX/NEON/etc parallel units that are often leveraged by ML libs, and there are FLOPS from integrated GPUs.  

The advertising of each is not consistent, and they tend to focus on the bigger number, often the iGPU but that is most likely not usable for ML in most AMD or Intel cases.

**Review by julien-c — COMMENTED** [review]
thanks a lot @apolinario!

**Comment by julien-c** [comment]
cc @mfuntowicz too on the whole thread btw

**Comment by julien-c** [comment]
@rwightman even if not cleanly comparable to GPUs, are the current CPU tflops values in the current state of this PR reasonable enough?

**Comment by rwightman** [comment]
@julien-c bit late to the party it seems, checking Ice Lake and Sapphire Rapids, they're both lower than I'd expect .. think Ice lake is in the 1-3 tflop range ans Sapphire is 3-4+..

Also, GPUs still way off, consumer cards about 1/2 what they should be and A100/H100 off by much more...

**Comment by julien-c** [comment]
@rwightman oops, PR welcome on top of this!

**Pull request merged**